### PR TITLE
All per-CPU structures should be allocated on EBPF_CACHE_LINE_SIZE boundaries

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Signed-off-by: Alan Jowett <alanjo@microsoft.com>

### CODE REVIEW

File: libs/platform/ebpf_platform.h [hunk 1]
```diff
@@ -26,6 +26,23 @@ extern "C"
         ((uint8_t*)(x)), sizeof((x)) - 1      \
     }
 
+#define EBPF_CACHE_LINE_SIZE (64)
```
Comment (dthaler) [inline libs/platform/ebpf_platform.h:29]:
```suggestion
#define EBPF_CACHE_LINE_SIZE 64
```

File: libs/platform/ebpf_platform.h [hunk 2]
```diff
@@ -26,6 +26,23 @@ extern "C"
...
+#define EBPF_CACHE_ALIGN_POINTER(P) (void*)(((uintptr_t)P + EBPF_CACHE_LINE_SIZE - 1) & ~(EBPF_CACHE_LINE_SIZE - 1))
+
+#define EBPF_DECLARE_ALIGNED_POINTER(TYPE, NAME)        \
```
Comment (dthaler) [inline libs/platform/ebpf_platform.h:32]: since they're declared as static below
```suggestion
#define EBPF_DECLARE_STATIC_ALIGNED_POINTER(TYPE, NAME)        \
```

**Review by dthaler — APPROVED** [review]

**Comment by Alan-Jowett** [comment]
Looks like ExAllocate with [POOL_FLAG_CACHE_ALIGNED](https://docs.microsoft.com/en-us/windows-hardware/drivers/kernel/pool_flags) and [_aligned_malloc](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/aligned-malloc?view=msvc-160) can significantly simplify this PR.

**Review by dthaler — APPROVED** [review]

**Comment by Alan-Jowett** [comment]
After switching to _aligned_alloc:

## Baseline
Test/Mode | Average   of Time | StdDev   of Time
-- | -- | --
test_program_invoke_interpret | 264 | 78
0 | 224 | 76
1 | 305 | 57
test_program_invoke_jit | 279 | 128
0 | 187 | 57
1 | 372 | 110

## This change

Test/Mode | Average   of Time | StdDev   of Time
-- | -- | --
test_program_invoke_interpret | 210 | 50
0 | 163 | 4
1 | 257 | 19
test_program_invoke_jit | 215 | 73
0 | 147 | 4
1 | 283 | 33

File: libs/platform/kernel/ebpf_platform_kernel.c [hunk 3]
```diff
@@ -47,6 +47,19 @@ ebpf_free(_Frees_ptr_opt_ void* memory)
...
 }
 
+__drv_allocatesMem(Mem) _Must_inspect_result_ _Ret_maybenull_
+    _Post_writable_byte_size_(size) void* ebpf_allocate_cache_aligned(size_t size)
+{
+    return ExAllocatePool2(POOL_FLAG_NON_PAGED | POOL_FLAG_CACHE_ALIGNED, size, EBPF_POOL_TAG);
+}
+
+void
+ebpf_free_cache_aligned(_Frees_ptr_opt_ void* memory)
```
Comment (shankarseal) [inline libs/platform/kernel/ebpf_platform_kernel.c:57]: why is this needed if its just a wrapper over ebpf_free?

**Review by shankarseal — APPROVED** [review]

File: libs/platform/kernel/ebpf_platform_kernel.c (hunk 3 above)
Comment (Alan-Jowett) [reply libs/platform/kernel/ebpf_platform_kernel.c:57]: For user mode it redirects to _alligned_free
